src/rust/engine/fs/store/src/lib.rs

```diff
@@ -951,55 +965,70 @@ impl Store {
     Ok(missing.is_empty())
   }
 
-  ///
-  /// Ensure that a directory is locally loadable, which will download it from the Remote store as
-  /// a sideeffect (if one is configured).
-  ///
-  pub async fn ensure_local_has_recursive_directory(
+  /// Ensure that the files are locally loadable. This will download them from the remote store as
+  /// a side effect, if one is configured.
+  pub async fn ensure_local_has_files(
```

Comment:
Maybe rename to: `ensure_local_has_digests`, `ensure_local_has`, `ensure_downloaded`, etc.

Somewhat implicit in here is that we also want to guarantee that the `Directory` protos have been persisted locally... so this method should probably also call `record_digest_trie` on each loaded tree to ensure that they exist locally as well. It's likely a bug that we weren't doing that before.

Reply:
Is the idea that at the end of callling `download_digest_to_local` on every missing file digest, we then also use a `future::try_join_all` that calls `ensure_directory_digest_persisted` on every directory digest inputted?

Reply:
> we then also use a `future::try_join_all` that calls `ensure_directory_digest_persisted` on every directory digest inputted?

Yea, either at the beginning or the end is fine. It's also possible to use [`try_join!`](https://github.com/pantsbuild/pants/blob/c1514b73a6248a366228549c95bc1cc00fc2d93c/src/rust/engine/process_execution/src/local.rs#L533-L536) to do it concurrently.

But you should be able to use `record_digest_trie` rather than `ensure_directory_digest_persisted`, because you will have already called `load_digest_trie` at the beginning of the method.

Reply:
@stuhood I want to make sure I got it right why it's safe to use `record_digest_trie` before we hit the remote store. That function is only about storing the directory entries, which don't need to come from the remote? We only download file entries from the remote?

Reply:
> That function is only about storing the directory entries, which don't need to come from the remote? We only download file entries from the remote?

Right: `record_digest_trie` only stores the `Directory` entries. Unlike files, `Directory` entries may be in one of three places: 1. memory, 2. local, 3. remote... `record_digest_trie` ensures that they are actually on disk locally, rather than potentially only in memory or remote.